parser.cc

```diff
@@ -1378,6 +1378,22 @@ char32_t Input::slowCharacterLookup(Index n)
 	return buffer[n - buffer_start];
 }
 Input::~Input() {}
+
+class NoInput : public Input
+{
+public:
+	NoInput() : Input("<none>") {}
+
+	bool fillBuffer(Index, Index&, char32_t*&) override { return false; }
+	Index size() const override { return 0; }
+};
+
+const Input& Input::iterator::input() const
+{
+	static NoInput None;
+	return buffer ? *buffer : None;
```

Comment:
I think it would be cleaner to expose a `filename()` method on the iterator and have it either call `input->filename()` or return the empty string (or possibly something like `"<invalid input>"`), since we don't really want people mapping back from iterators to inputs in the general case.